src/main/java/com/alvarium/annotators/PkiHttpAnnotator.java

```diff
@@ -0,0 +1,146 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package com.alvarium.annotators;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.URISyntaxException;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.nio.file.Path;
+
+import com.alvarium.annotators.http.ParseResult;
+import com.alvarium.annotators.http.ParseResultException;
+import com.alvarium.contracts.Annotation;
+import com.alvarium.contracts.AnnotationType;
+import com.alvarium.hash.HashType;
+import com.alvarium.sign.KeyInfo;
+import com.alvarium.sign.SignException;
+import com.alvarium.sign.SignProvider;
+import com.alvarium.sign.SignProviderFactory;
+import com.alvarium.sign.SignatureInfo;
+import com.alvarium.sign.SignType;
+import com.alvarium.utils.Encoder;
+import com.alvarium.utils.PropertyBag;
+
+import org.apache.http.client.methods.HttpUriRequest;
+
+class PkiHttpAnnotator extends AbstractAnnotator implements Annotator {
+  private final HashType hash;
+  private SignatureInfo signature;
+  private final AnnotationType kind;
+
+  protected PkiHttpAnnotator(HashType hash, SignatureInfo signature) {
+    this.hash = hash;
+    this.signature = signature;
+    this.kind = AnnotationType.PKIHttp;
+  }
+
+  public Annotation execute(PropertyBag ctx, byte[] data) throws AnnotatorException {
+    final String key = super.deriveHash(hash, data);
+
+    String host;
+    try {
+      host = InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      throw new AnnotatorException("Cannot get host name", e);
+    }
+
+    HttpUriRequest request = ctx.getProperty(AnnotationType.PKIHttp.name(), HttpUriRequest.class);
+    ParseResult parsed; 
+    try {
+      parsed = new ParseResult(request);
+    } catch (URISyntaxException e) {
+      throw new AnnotatorException("Invalid request URI", e);
+    } catch (ParseResultException e) {
+      throw new AnnotatorException("Error parsing the request", e);
+    }
+    final Signable signable = new Signable(parsed.getSeed(), parsed.getSignature());
+
+    // Use the parsed request to obtain the key name and type we should use to
+    // validate the signature
+    Path path = Paths.get(signature.getPublicKey().getPath());
+    Path directory = path.getParent();
+    String publicKeyPath = String.join("/", directory.toString(), parsed.getKeyid());
+
+    SignType alg;
+    try {
+      alg = SignType.fromString(parsed.getAlgorithm());
+    } catch (EnumConstantNotPresentException e) {
+      throw new AnnotatorException("Invalid key type " + parsed.getAlgorithm());
+    }
+    KeyInfo publicKey = new KeyInfo(publicKeyPath, alg);
+    signature = new SignatureInfo(publicKey, signature.getPrivateKey());
```

Comment:
Setting the `SignatureInfo` that way would result in a mutable state inside the `Annotator` which will change in every `execute` method call, leading to unexpected behavior with each consecutive call to the method. Also, overwriting the previously passed `SignatureInfo` will result in the loss of the original config, which may come in handy incase of fallbacks. I believe it is not the responsibility of the `execute` method to set the `Annotator` config explicitly. I suggest changing the `signature` field to be `final` and create a new local variable which will hold the newly created `SignatureInfo` to achieve immutability and preserve the original config.